### PR TITLE
[SPARK-22566][PYTHON] Better error message for `_merge_type` in Pandas to Spark DF conversion

### PR DESCRIPTION
**Pull request by gberger**
## What changes were proposed in this pull request?

It provides a better error message when doing `spark_session.createDataFrame(pandas_df)` with no schema and an error occurs in the schema inference due to incompatible types.

The Pandas column names are propagated down and the error message mentions which column had the merging error. 

https://issues.apache.org/jira/browse/SPARK-22566

## How was this patch tested?

Manually in the `./bin/pyspark` console, and with new tests: `./python/run-tests`

screen shot 2017-11-21 at 13 29 49

I state that the contribution is my original work and that I license the work to the Apache Spark project under the project’s open source license.

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -1108,19 +1109,22 @@ def _has_nulltype(dt):
...
-def _merge_type(a, b):
+def _merge_type(a, b, name=None):
     if isinstance(a, NullType):
         return b
     elif isinstance(b, NullType):
         return a
     elif type(a) is not type(b):
         # TODO: type cast (such as int -> long)
-        raise TypeError("Can not merge type %s and %s" % (type(a), type(b)))
+        if name is not None:
```
Comment (ebuildy) [inline python/pyspark/sql/types.py]: Easier to read as:

```
if name is None:
    raise TypeError("Can not merge type %s and %s" % (type(a), type(b)))
else:
    raise TypeError("Can not merge type %s and %s in column %s" % (type(a), type(b), name))
```

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -1108,19 +1109,22 @@ def _has_nulltype(dt):
         return isinstance(dt, NullType)
 
 
-def _merge_type(a, b):
+def _merge_type(a, b, name=None):
```
Comment (ueshin) [inline python/pyspark/sql/types.py]: When does this take the `name` parameter?

**Comment by ueshin** [comment]
What if `ArrayType` or `MapType`, or deeply nested `StructType`?

**Comment by gberger** [comment]
@ebuildy fixed

**Comment by gberger** [comment]
@ueshin 

The reason that I modified the case for StructType is that, in session.py#341, for each Pandas DF row we obtain a StructType with StructFields mapping column names to value type; these are reduced with _merge_types.

I do appreciate that it could be the case that a Pandas DF contains lists or dicts as values. I pushed a new commit where the `name` property gets passed down when we are recursing via ArrayType or MapType.

Here is what it looks like when we use lists or dicts inside a Pandas DF:

```
>>> df = pd.DataFrame(data={
...     'a': [[1, 2], [3, 4]],
...     'b': [[5, 'hello'], [7, 8]]
... })
>>> sdf = sql.createDataFrame(df)
>>> sdf
DataFrame[a: array<bigint>, b: array<bigint>]
>>> sdf.show()
+------+---------+
|     a|        b|
+------+---------+
|[1, 2]|[5, null]|
|[3, 4]|   [7, 8]|
+------+---------+
```

```
>>> df = pd.DataFrame(data={
...     'a': [{1: 2}, {3: 4}],
...     'b': [{5: 'hello'}, {7: 8}]
... })
>>> sdf = sql.createDataFrame(df)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gberger/Projects/spark/python/pyspark/sql/context.py", line 354, in createDataFrame
    return self.sparkSession.createDataFrame(data, schema, samplingRatio, verifySchema)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 646, in createDataFrame
    rdd, schema = self._createFromLocal(map(prepare, data), schema)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 409, in _createFromLocal
    struct = self._inferSchemaFromList(data, names=schema)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 341, in _inferSchemaFromList
    schema = reduce(_merge_type, [_infer_schema(row, names) for row in data])
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1128, in _merge_type
    for f in a.fields]
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1128, in <listcomp>
    for f in a.fields]
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1140, in _merge_type
    _merge_type(a.valueType, b.valueType, name=name),
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1122, in _merge_type
    raise TypeError("Can not merge type %s and %s in field '%s'" % (type(a), type(b), name))
TypeError: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.LongType'> in field b
```

**Comment by gberger** [comment]
Jenkins, please retest

**Comment by gberger** [comment]
Jenkins, retest this please

**Comment by gberger** [comment]
@ueshin I think this build fail was an outage, can we retest?

    Could not load hsdis-amd64.so; library not loadable; PrintAssembly is disabled

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by ueshin** [comment]
Can you add tests for this?
Maybe we need some tests for simple case and some nested type cases.
I'd also cc @HyukjinKwon. Thanks.

**Comment by HyukjinKwon** [comment]
Thanks @ueshin. Yup, +1 for adding some tests.

I just wonder if we could have a similar form of error message in type verification. I remember I fixed a similar issue for type verification - https://github.com/apache/spark/pull/18521 (see the links in "Before" and "After"), for example:

```
field c in field b in field a: can not merge type IntegerType and StringType
element in array element in array field a: can not merge type IntegerType and StringType
```

Let's make sure there is no performance regression as well (even I was about to make the mistake before).

**Comment by HyukjinKwon** [comment]
@gberger, BTW, just to be clear, IIRC the type inference and merging code path here are shared for other data types, for example, dict, namedtuple, row and etc.

**Comment by gberger** [comment]
For sure @ueshin, I will add tests.

@HyukjinKwon understood! 
How do we go about testing performance regression?

**Comment by HyukjinKwon** [comment]
I think you can add a new test case somewhere around here - https://github.com/apache/spark/blob/master/python/pyspark/sql/tests.py#L1724 maybe dealing with some combinations of `StructType`, `MapType`, `ArrayType` and one of non-nested type (e.g., `IntegerType`), checking if the error message is really improved.

For running tests, you could refer https://spark.apache.org/docs/latest/building-spark.html#pyspark-tests-with-maven.

I personally test them locally first and then add and run some tests. Maybe running `./python/run-tests --modules=pyspark-sql` could be enough. Also, make sure running `./dev/lint-python` to check if it complies PEP8 and etc.

When you finish them, commit and push which will trigger the build via Jenkins here. See also "Pull Request" in http://spark.apache.org/contributing.html

**Comment by HyukjinKwon** [comment]
D'oh, you mean performance regression test. Manual tests should be fine. When you share some codes you ran, maybe we can double check.

**Comment by gberger** [comment]
Hey all,

## Error message

I revamped the error message and made it "recursive" similar to @HyukjinKwon. Here's an example:

```
>>> _merge_type(
...                 StructType([StructField("f1", ArrayType(MapType(StringType(), LongType())))]),
...                 StructType([StructField("f1", ArrayType(MapType(DoubleType(), LongType())))])
...             )
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1129, in _merge_type
    for f in a.fields]
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1129, in <listcomp>
    for f in a.fields]
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1137, in _merge_type
    return ArrayType(_merge_type(a.elementType, b.elementType, field=field+'.arrayElement'), True)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1140, in _merge_type
    return MapType(_merge_type(a.keyType, b.keyType, field=field+'.mapKey'),
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1122, in _merge_type
    raise TypeError("%s: Can not merge type %s and %s" % (field, type(a), type(b)))
TypeError: .structField("f1").arrayElement.mapKey: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.DoubleType'>
```

Happy to iterate on the exact formatting or wording of the path shown.

## Tests

I wrote a bunch of tests too, hope they are comprehensive enough but happy to add more if not. @ueshin 


## Benchmark

It seems that the time it takes for a nested _merge_type on my machine has increased for ~2.75 microseconds to  ~2.85 microseconds, around a 3% increase. 

This can be attributed to the string concatenation that goes on every time _merge_type goes one level down from a StructType, ArrayType or MapType. I'm not sure if there's a better way to propagate this information down the stack, maybe a tuple?

Code used:

```
from pyspark.sql.types import *
from pyspark.sql.types import _merge_type
import time


def test_f():
    return _merge_type(
        StructType([StructField("f1", ArrayType(MapType(StringType(), LongType())))]),
        StructType([StructField("f1", ArrayType(MapType(StringType(), LongType())))])
    )

def timing(f):
    def wrap(*args):
        time1 = time.time()
        for __ in range(100000):
            ret = f(*args)
        time2 = time.time()
        print('took %0.3f ms' % ((time2-time1)*1000.0))
        return ret
    return wrap

for _ in range(10):
    timing(test_f)()

```
Before:

> took 2701.337 ms
> took 2905.867 ms
> took 2725.119 ms
> took 2796.098 ms
> took 2718.981 ms
> took 2773.560 ms
> took 2717.995 ms
> took 2796.466 ms
> took 2716.173 ms
> took 2744.121 ms


After:

> took 2865.038 ms
> took 2836.403 ms
> took 2871.871 ms
> took 2827.625 ms
> took 2820.170 ms
> took 2873.976 ms
> took 2833.609 ms
> took 2909.599 ms
> took 3162.108 ms
> took 2940.864 ms

**Comment by gberger** [comment]
Maybe a more performant way to do the path in the error message would be to propagate it *up* the stack via try/catching the errors and adding the paths as it goes.

But this way seems really weird to me...

```
    elif isinstance(a, ArrayType):
        try: 
            return ArrayType(_merge_type(a.elementType, b.elementType), True)
        catch TypeError as e:
            raise TypeError("arrayElement." + str(e))
```

File: python/pyspark/sql/types.py [hunk 3]
```diff
@@ -1108,19 +1109,23 @@ def _has_nulltype(dt):
         return isinstance(dt, NullType)
 
 
-def _merge_type(a, b):
+def _merge_type(a, b, path=''):
```
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/sql/types.py]: Yup, I agree with it for the current status .. 

it's kind of difficult to come up with a good message format in such cases. To me, I actually kind of gave up a pretty format and just chose prose before (in the PR #18521).

I still don't have a good idea to show the nested structure in the error message to be honest. I was thinking one of prose, kind of piece of codes (like `schema['f1'].dataType.elementType.keyType` instead of https://github.com/apache/spark/pull/19792#issuecomment-346681841), or somehow pretty one like `printSchema()` ... ?

Maybe, I am thinking of referring other formats in Spark or somewhere like Pandas or piece of codes for now.

So, to cut it short, here are what are on my mind for the example of https://github.com/apache/spark/pull/19792#issuecomment-346681841:

1. Prose

```
TypeError: key in map field in array element in field f1: Can not blabla
```

2. Piece of codes with `StructType`

```
TypeError: schema(?)['f1'].dataType.elementType.keyType: Can not blabla
TypeError: struct(?)['f1'].dataType.elementType.keyType: Can not blabla
```


3. `printSchema()`

```
TypeError: 
root
 |-- f1: array (nullable = false)
 |    |-- element: map (containsNull = false)
 |    |    |-- key: integer*
 |    |    |-- value: long (valueContainsNull = false)
: *Can not blabla
```

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/sql/types.py]: I really don't have a strong opinion on this. @ueshin do you maybe have a preference or another option, or are you okay with the current format in general?

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/sql/types.py]: Maybe, just printing the string in SQL could be an option too to be consistent .. it's a bit messy for deeply nested schema though.:

```
TypeError: struct<f1:array<map<int,string>>> and struct<f1:array<map<int,int>>> Can not blabla
```

[the remark this answers is not in the document]
Comment (ueshin) [reply python/pyspark/sql/types.py]: I prefer the format we can know the path to the error and I guess we don't need the pretty one like the `printSchema()`.
Maybe following the #18521 format is good enough, and the current one is good as well but I wanted something in front of the path string instead of starting with `.`.

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/sql/types.py]: Yup, looks we need something ahead anyway. To me, I am okay with any format. 

I think we could fix later again. Just I wondered if you have a preference.

File: python/pyspark/sql/tests.py [hunk 4]
```diff
@@ -1722,6 +1723,83 @@ def test_infer_long_type(self):
...
+        with self.assertRaisesRegexp(TypeError, 'mapValue'):
+            _merge_type(
+                MapType(StringType(), LongType()),
+                MapType(StringType(), DoubleType()))
+
+        self.assertEqual(_merge_type(
+            StructType([StructField("f1", LongType()), StructField("f2", StringType())]),
+            StructType([StructField("f1", LongType()), StructField("f2", StringType())])
+        ), StructType([StructField("f1", LongType()), StructField("f2", StringType())]))
+        with self.assertRaisesRegexp(TypeError, r'structField\("f1"\)'):
```
[the remark this answers is not in the document]
Comment (gberger) [reply python/pyspark/sql/tests.py]: But then double backslashes? :(

[the remark this answers is not in the document]
Comment (ueshin) [reply python/pyspark/sql/tests.py]: Hmm, I don't think we need double backslashes here.

[the remark this answers is not in the document]
Comment (gberger) [reply python/pyspark/sql/tests.py]: Oh, you are right. Fixed

File: python/pyspark/sql/types.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gberger) [reply python/pyspark/sql/types.py]: Hi folks, 

Great options proposed by @HyukjinKwon - though I did not comprehend what was the conclusion of the discussion with @ueshin :)

Which format should we employ? And do you want me to use this format right now or is it something you'll fix later?

Thanks!

[the remark this answers is not in the document]
Comment (ueshin) [reply python/pyspark/sql/types.py]: Well, can you follow the #18521 format for now? Thanks!

**Comment by gberger** [comment]
Hi all,

I have changed the error message to be like #18521. Here are some examples:

```
_merge_type(
    StructType([StructField("f1", ArrayType(MapType(StringType(), LongType())))]),
    StructType([StructField("f1", ArrayType(MapType(DoubleType(), LongType())))])
)
# TypeError: key of map element in array field f1: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.DoubleType'>

_merge_type(
    StructType([
        StructField("f1", MapType(StringType(), LongType())),
        StructField("f2", StringType())]),
    StructType([
        StructField("f1", MapType(StringType(), DoubleType())),
        StructField("f2", StringType())]))
# TypeError: value of map field f1: Can not merge type <class 'pyspark.sql.types.LongType'> and <class 'pyspark.sql.types.DoubleType'>

_merge_type(
    StructType([
        StructField("f1", ArrayType(LongType())),
        StructField("f2", StringType())]),
    StructType([
        StructField("f1", ArrayType(DoubleType())),
        StructField("f2", StringType())]))
# TypeError: element in array field f1: Can not merge type <class 'pyspark.sql.types.LongType'> and <class 'pyspark.sql.types.DoubleType'>

_merge_type(
    StructType([StructField("f1", LongType()), StructField("f2", StringType())]),
    StructType([StructField("f1", DoubleType()), StructField("f2", StringType())]))
# TypeError: field f1: Can not merge type <class 'pyspark.sql.types.LongType'> and <class 'pyspark.sql.types.DoubleType'>

_merge_type(
    StructType([StructField("f1", StructType([StructField("f2", LongType())]))]),
    StructType([StructField("f1", StructType([StructField("f2", StringType())]))]))
# TypeError: field f2 in field f1: Can not merge type <class 'pyspark.sql.types.LongType'> and <class 'pyspark.sql.types.StringType'>
```

File: python/pyspark/sql/types.py [hunk 5]
```diff
@@ -1108,19 +1109,33 @@ def _has_nulltype(dt):
         return isinstance(dt, NullType)
 
 
-def _merge_type(a, b):
+def _merge_type_path(path, addition):
```
Comment (ueshin) [inline python/pyspark/sql/types.py]: This is not used?

**Review by ueshin — COMMENTED** [review]
LGTM except for a minor comment.

File: python/pyspark/sql/types.py (hunk 5 above)
Comment (gberger) [reply python/pyspark/sql/types.py]: oopsie

**Comment by gberger** [comment]
Friendly ping -- I've fixed that @ueshin.
Is there anything else I should look at to get this to be merged?
/cc @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Oh, will double check too for sure shortly.

File: python/pyspark/sql/session.py [hunk 6]
```diff
@@ -337,7 +338,7 @@ def _inferSchemaFromList(self, data):
         if type(first) is dict:
             warnings.warn("inferring schema from dict is deprecated,"
                           "please use pyspark.sql.Row instead")
-        schema = reduce(_merge_type, map(_infer_schema, data))
+        schema = reduce(_merge_type, [_infer_schema(row, names) for row in data])
```
Comment (HyukjinKwon) [inline python/pyspark/sql/session.py]: Not a big deal but let's use generator expression -> `(_infer_schema(row, names) for row in data)`

**Review by HyukjinKwon — APPROVED** [review]
LGTM too.

**Comment by gberger** [comment]
Fixed

**Comment by HyukjinKwon** [comment]
Can we set the `names` for `_createFromRDD` -> `_inferSchema` too?:

```python
>>> spark.createDataFrame(spark.sparkContext.parallelize([[None, 1], ["a", None], [1, 1]]), schema=["a", "b"], samplingRatio=0.99)
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/hyukjinkwon/Desktop/workspace/repos/forked/spark/python/pyspark/sql/session.py", line 644, in createDataFrame
    rdd, schema = self._createFromRDD(data.map(prepare), schema, samplingRatio)
  File "/Users/hyukjinkwon/Desktop/workspace/repos/forked/spark/python/pyspark/sql/session.py", line 383, in _createFromRDD
    struct = self._inferSchema(rdd, samplingRatio)
  File "/Users/hyukjinkwon/Desktop/workspace/repos/forked/spark/python/pyspark/sql/session.py", line 375, in _inferSchema
    schema = rdd.map(_infer_schema).reduce(_merge_type)
  File "/.../spark/python/pyspark/rdd.py", line 852, in reduce
    return reduce(f, vals)
  File "/.../spark/python/pyspark/sql/types.py", line 1133, in _merge_type
    for f in a.fields]
  File "/.../spark/python/pyspark/sql/types.py", line 1126, in _merge_type
    raise TypeError(new_msg("Can not merge type %s and %s" % (type(a), type(b))))
TypeError: field _1: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.LongType'>
```

File: python/pyspark/sql/session.py [hunk 7]
```diff
@@ -324,11 +324,12 @@ def range(self, start, end=None, step=1, numPartitions=None):
 
         return DataFrame(jdf, self._wrapped)
 
-    def _inferSchemaFromList(self, data):
+    def _inferSchemaFromList(self, data, names=None):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/session.py]: Can we set the `names` for `_createFromRDD` -> `_inferSchema` too?:

```python
>>> spark.createDataFrame(spark.sparkContext.parallelize([[None, 1], ["a", None], [1, 1]]), schema=["a", "b"], samplingRatio=0.99)
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../spark/python/pyspark/sql/session.py", line 644, in createDataFrame
    rdd, schema = self._createFromRDD(data.map(prepare), schema, samplingRatio)
  File "/.../spark/python/pyspark/sql/session.py", line 383, in _createFromRDD
    struct = self._inferSchema(rdd, samplingRatio)
  File "/.../spark/python/pyspark/sql/session.py", line 375, in _inferSchema
    schema = rdd.map(_infer_schema).reduce(_merge_type)
  File "/.../spark/python/pyspark/rdd.py", line 852, in reduce
    return reduce(f, vals)
  File "/.../spark/python/pyspark/sql/types.py", line 1133, in _merge_type
    for f in a.fields]
  File "/.../spark/python/pyspark/sql/types.py", line 1126, in _merge_type
    raise TypeError(new_msg("Can not merge type %s and %s" % (type(a), type(b))))
TypeError: field _1: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.LongType'>
```

**Comment by gberger** [comment]
@HyukjinKwon done, with test added. 

```
>>> spark.createDataFrame(spark.sparkContext.parallelize([[None, 1], ["a", None], [1, 1]]), schema=["a", "b"], samplingRatio=0.99)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 644, in createDataFrame
    rdd, schema = self._createFromRDD(data.map(prepare), schema, samplingRatio)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 383, in _createFromRDD
    struct = self._inferSchema(rdd, samplingRatio, names=schema)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 375, in _inferSchema
    schema = rdd.map(lambda row: _infer_schema(row, names)).reduce(_merge_type)
  File "/Users/gberger/Projects/spark/python/pyspark/rdd.py", line 852, in reduce
    return reduce(f, vals)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1133, in _merge_type
    for f in a.fields]
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1126, in _merge_type
    raise TypeError(new_msg("Can not merge type %s and %s" % (type(a), type(b))))
TypeError: field a: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.LongType'>
```

Also, with this last change, I could simplify the code in `_createFromRDD`. Since I pass the field names down to `_inferSchema` (and to `_infer_schema` from there), the inferred schema already comes with field names, so no need to set them again in `_createFromRDD`. Tests for this still pass. Let me know if you can think of any edge case not covered by tests that would break.

File: python/pyspark/sql/session.py [hunk 8]
```diff
@@ -405,7 +401,7 @@ def _createFromLocal(self, data, schema):
             data = list(data)
 
         if schema is None or isinstance(schema, (list, tuple)):
-            struct = self._inferSchemaFromList(data)
+            struct = self._inferSchemaFromList(data, names=schema)
             converter = _create_converter(struct)
             data = map(converter, data)
             if isinstance(schema, (list, tuple)):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/session.py]: @gberger, could we remove this branch too?

Comment (gberger) [reply python/pyspark/sql/session.py]: yes

Comment (gberger) [reply python/pyspark/sql/session.py]: removed with commit 5131db2

File: python/pyspark/sql/session.py [hunk 9]
```diff
@@ -405,14 +401,9 @@ def _createFromLocal(self, data, schema):
             data = list(data)
 
         if schema is None or isinstance(schema, (list, tuple)):
-            struct = self._inferSchemaFromList(data)
-            converter = _create_converter(struct)
+            schema = self._inferSchemaFromList(data, names=schema)
+            converter = _create_converter(schema)
             data = map(converter, data)
-            if isinstance(schema, (list, tuple)):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/session.py]:
```python
>>> spark.createDataFrame([{'a': 1}], ["b"])
DataFrame[a: bigint]
```

Hm.. sorry actually I think we should not remove this one and L385 because we should primarily respect user's input and it should be `DataFrame[b: bigint]`.

**Review by HyukjinKwon — COMMENTED** [review]
Latest changes LGTM except for one comment.

**Comment by gberger** [comment]
Good catch @HyukjinKwon! I reverted those changes and added a test to cover this regression.

File: python/pyspark/sql/types.py [hunk 10]
```diff
@@ -1083,7 +1083,8 @@ def _infer_schema(row):
         elif hasattr(row, "_fields"):  # namedtuple
             items = zip(row._fields, tuple(row))
         else:
-            names = ['_%d' % i for i in range(1, len(row) + 1)]
+            if names is None:
+                names = ['_%d' % i for i in range(1, len(row) + 1)]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py]: @gberger, Let's revert this change too. Seems it's going to introduce a behaviour change:

**Before**

```
>>> spark.createDataFrame([["a", "b"]], ["col1"]).show()
17/12/12 15:29:45 WARN ObjectStore: Failed to get database global_temp, returning NoSuchObjectException
+----+---+
|col1| _2|
+----+---+
|   a|  b|
+----+---+
```

**After**
```
>>> spark.createDataFrame([["a", "b"]], ["col1"]).show()
...
java.lang.IllegalStateException: Input row doesn't have expected number of values required by the schema. 1 fields are required while 2 values are provided.
	at org.apache.spark.sql.execution.python.EvaluatePython$.fromJava(EvaluatePython.scala:148)
        ...
```

Comment (gberger) [reply python/pyspark/sql/types.py]: You're right, but by reverting we lose the nice message. Notice below reverted it says field _1, where it could have said field col1.

```
>>> spark.createDataFrame([["a", "b"], [1, 2]], ["col1"]).show()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 646, in createDataFrame
    rdd, schema = self._createFromLocal(map(prepare, data), schema)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 409, in _createFromLocal
    struct = self._inferSchemaFromList(data, names=schema)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 341, in _inferSchemaFromList
    schema = reduce(_merge_type, (_infer_schema(row, names) for row in data))
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1132, in _merge_type
    for f in a.fields]
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1125, in _merge_type
    raise TypeError(new_msg("Can not merge type %s and %s" % (type(a), type(b))))
TypeError: field _1: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.LongType'>
```

Instead, I am adding a new elif branch where we check len(names) vs len(row). If we have fewer names than we have columns, we extend the names list, completing it with entries such as _2.

```
>>> spark.createDataFrame([["a", "b"]], ["col1"])
DataFrame[col1: string, _2: string]
```

I have included a test for this as well.

Comment (gberger) [reply python/pyspark/sql/types.py]: If we revert it, then the original purpose of the PR is lost:

```
>>> df = pd.DataFrame(data={'a':[1,2,3], 'b': [4, 5, 'hello']})
>>> spark.createDataFrame(df)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 646, in createDataFrame
    rdd, schema = self._createFromLocal(map(prepare, data), schema)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 409, in _createFromLocal
    struct = self._inferSchemaFromList(data, names=schema)
  File "/Users/gberger/Projects/spark/python/pyspark/sql/session.py", line 341, in _inferSchemaFromList
    schema = reduce(_merge_type, (_infer_schema(row, names) for row in data))
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1132, in _merge_type
    for f in a.fields]
  File "/Users/gberger/Projects/spark/python/pyspark/sql/types.py", line 1125, in _merge_type
    raise TypeError(new_msg("Can not merge type %s and %s" % (type(a), type(b))))
TypeError: field _2: Can not merge type <class 'pyspark.sql.types.LongType'> and <class 'pyspark.sql.types.StringType'>
```

Comment (gberger) [reply python/pyspark/sql/types.py]: @HyukjinKwon just pushed the elif branch change that I talked about above, please see if it is suitable

Comment (HyukjinKwon) [reply python/pyspark/sql/types.py]: Yup, will take another look soon anyway.

File: python/pyspark/sql/types.py [hunk 11]
```diff
@@ -1083,7 +1083,11 @@ def _infer_schema(row):
...
+            elif len(names) < len(row):
+                names = names[:]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py]: Hm .. why we do this? to copy?

Comment (gberger) [reply python/pyspark/sql/types.py]: Yes, I did not want to modify the original list since `.extend` is an in-place operation. However, session.py#602 already creates a copy of the list passed by the user, so this copying in `_infer_schema` is actually not necessary. Removing now.

**Comment by HyukjinKwon** [comment]
Will take another look soon tomorrow. Sorry that is getting delayed again and again but I just realised this code path is a little bit tricky ..

**Comment by gberger** [comment]
@HyukjinKwon no worries, I understand. We gotta be 100% thorough here. Thanks for the help

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by HyukjinKwon** [comment]
To be clear (as I was reviewing this too), I am okay with going ahead, @ueshin if this looks good to you.

**Comment by ueshin** [comment]
@HyukjinKwon Thanks, I'll take another look soon.

**Comment by ueshin** [comment]
LGTM, pending Jenkins.

**Comment by ueshin** [comment]
Thanks! merging to master/2.3.

**Comment by gberger** [comment]
Great! Thanks all
